src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeQueries.java

```diff
@@ -0,0 +1,181 @@
+    {
+        @Override
+        public boolean visit(byte[] packedValue)
+        {
+            return true;
+        }
+
+        @Override
+        public Relation compare(byte[] minPackedValue, byte[] maxPackedValue)
+        {
+            return Relation.CELL_INSIDE_QUERY;
+        }
+    };
+
+    public static BlockBalancedTreeReader.IntersectVisitor balancedTreeQueryFrom(Expression expression, int bytesPerValue)
+    {
+        if (expression.lower == null && expression.upper == null)
+        {
+            return MATCH_ALL;
+        }
+
+        Bound lower = null ;
+        if (expression.lower != null)
+        {
+            final byte[] lowerBound = toComparableBytes(bytesPerValue, expression.lower.value.encoded, expression.validator);
+            lower = new Bound(lowerBound, !expression.lower.inclusive);
+        }
+
+        Bound upper = null;
+        if (expression.upper != null)
+        {
+            final byte[] upperBound = toComparableBytes(bytesPerValue, expression.upper.value.encoded, expression.validator);
+            upper = new Bound(upperBound, !expression.upper.inclusive);
+        }
+
+        return new RangeQueryVisitor(bytesPerValue, lower, upper);
+    }
+
+    private static byte[] toComparableBytes(int bytesPerDim, ByteBuffer value, AbstractType<?> type)
+    {
+        byte[] buffer = new byte[TypeUtil.fixedSizeOf(type)];
+        assert buffer.length == bytesPerDim;
+        TypeUtil.toComparableBytes(value, type, buffer);
+        return buffer;
+    }
+
+    private static abstract class RangeQuery implements BlockBalancedTreeReader.IntersectVisitor
+    {
+        final int bytesPerValue;
+
+        RangeQuery(int bytesPerValue)
+        {
+            this.bytesPerValue = bytesPerValue;
+        }
+
+        int compareUnsigned(byte[] packedValue, Bound bound)
+        {
+            return FutureArrays.compareUnsigned(packedValue, 0, bytesPerValue, bound.bound, 0, bytesPerValue);
+        }
+    }
+
+    private static class Bound
+    {
+        private final byte[] bound;
+        private final boolean exclusive;
+
+        Bound(byte[] bound, boolean exclusive)
+        {
+            this.bound = bound;
+            this.exclusive = exclusive;
+        }
+
+        boolean smallerThan(int cmp)
+        {
+            return cmp > 0 || (cmp == 0 && exclusive);
+        }
+
+        boolean greaterThan(int cmp)
+        {
+            return cmp < 0 || (cmp == 0 && exclusive);
+        }
+    }
+
+    private static class RangeQueryVisitor extends RangeQuery
+    {
+        private final Bound lower;
+        private final Bound upper;
+
+        private RangeQueryVisitor(int bytesPerValue, Bound lower, Bound upper)
+        {
+            super(bytesPerValue);
+            this.lower = lower;
+            this.upper = upper;
+        }
+
+        @Override
+        public boolean visit(byte[] packedValue)
```

Comment:
nit: Part of me feels like this method would be better named `within()` or `contains()` or something less generic :/

Reply:
I've chosen `contains` because this is pretty much what it does.